hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMAllocateBlockRequest.java

```diff
@@ -192,6 +193,17 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
             KEY_NOT_FOUND);
       }
 
+      if (ozoneManager.getEnableFileSystemPaths()) {
+        // Ensure the parent exist.
+        if (!"".equals(OzoneFSUtils.getParent(keyName))
```

Comment:
Just a question: Right now delete/rename is not an atomic operation, so will just checking the parent Dir is enough?


But this will help, once we have HDDS-2939 where we have atomic delete/renames.

Reply:
And also for performance reasons, can we leave allocateBlock, as during KeyCommit anyway we shall figure it out?

As this will save one extra DB op, (this might not be a problem again once HDDS-2939 comes in, as it has directory cache)

Reply:
Thanx @bharatviswa504 for the review.
Rakesh suggested to add the check in `allocateBlock`, I think for a `fail-fast` approach, If we tend to fail during the commit phase and we know that at the `allocateBlock` phase, we should fail at an earlier stage itself, rather than going till the end. Might save a messed up client, if he is writing a big file.

But I am not very aware about the performance impacts and HDDS-2939 both, so I am ok both ways.

@rakeshadr any thoughts on this?

Reply:
>And also for performance reasons, can we leave allocateBlock, as during KeyCommit anyway we shall figure it out?

On a second thought, it makes sense to me to avoid any additional DB calls which can affect the OM performance. Most of the cases, parent directories would exists in RocksDB offheap cache even then I would agree to skip parentDir check considering the cost of DB call if not exist.

On the other side, I think there won't be any correctness issue if we skip parent dir check in AllocateBlockReq. Later, we can bring this` fast-fail` check based on user demand. Considering that, I'm +1 to skip extra DB check in AllocateBlockReq now.

Thanks a lot @ayushtkn for your patience and taking care this. I agree to revert the changes from OmAllocateBlockRequest.

>But I am not very aware about the performance impacts and HDDS-2939 both

In HDDS-2939, in order to fetch openFileName, it requires to traverse top-down model and parentDir check won't add any additional overhead. For example, say open file /a/b/c/d/file1.

_HDDS-2939 branch_: OM will fetch /a, /a/b, /a/b/c, /a/b/c/d from DirTable and finally get /a/b/c/d/file1 from OpenFileTable. Here, one idea is to build DirCache to reduce DB overhead. @bharatviswa504 was talking about this part in his previous comment.
_Master branch_: OM will do point look up by fetching /a/b/c/d from KeyTable. Here, existing RocksDB offheap cache would help.

Reply:
Thanx Rakesh for sharing the details, makes sense, Will explore more :-)

I have removed the check from `AllocateBlock` as suggested.

@bharatviswa504 Please give a check once agin, if things make sense now. Thanx...

Reply:
>Is there anything else, that can be checked? Yahh, there might be some issue because of non atomic behaviour delete/rename. Will pulling this check inside lock, solve them or lead to a better state?

As due to non-atomic behavior whole path check can help here. But this issue will be solved once HDDS-2939 comes in. And also this check is done under lock even right now. Let's have this way for now.